README.md
# CBT3.5 - ChatBotTurbo3.5 🚀

CBT3.5 is an amazing program that utilizes the Discord API and the power of ChatGPT 3.5 Turbo to create an interactive bot that can respond to user messages. 😮✨

## Prerequisites
Before you can run CBT3.5, make sure you have the following:

✅ Python installed on your system. ✔️

## Getting Started
1. Clone this repository to your local machine or download the source code. 📥

2. Install the required dependencies by running the following command:
pip install -r requirements.txt

3. Run CBT3.5 using the following command:
python3 main.py

4. The bot should now be online and ready to respond to user messages in the specified Discord server. 🤖💬

## Features
🔹 The bot listens for messages in a channel named "ask-anything" and responds using the power of ChatGPT 3.5 Turbo. 💡🎙️

🔹 Users with the "Special Access" and "Verified" role can ask questions and receive responses from the bot. 🙋‍♂️🤖

🔹 User questions and bot responses are automatically deleted after a certain period to keep the channel clean. 🗑️🧹


## Customization
You can customize the behavior of CBT3.5 by modifying the following variables in the program:

⚙️ command_prefix: The prefix used to invoke bot commands (default: !).

⚙️ intents: The Discord intents to enable for the bot (default: discord.Intents.default()).

⚙️ engine: The ChatGPT engine to use for generating responses (default: text-davinci-003).

⚙️ max_tokens: The maximum number of tokens in the generated response (default: 100).

⚙️ temperature: The temperature parameter for response generation (default: 0.7).

⚙️ delete_delay: The delay (in seconds) before deleting user messages and bot responses (default: 120 and 60 respectively).

Feel free to modify these variables according to your needs. 🔧🔩

## Contributions
🤝 Contributions are welcome! If you have any suggestions, improvements, or bug fixes, feel free to open an issue or submit a pull request.🐛🔧

## Credits
CBT3.5 is developed by GR1MR34P3R-1. 👏

## License
This program is licensed under the [MIT License](LICENSE). Feel free to use, modify, and distribute it as per the terms of the license.📄

## Have A Good Day!!
🌟 Have fun exploring the capabilities of CBT3.5 and enjoy interacting with the Discord bot! If you have any questions or need assistance, don't hesitate to reach out. 😊


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file.

## Released

### Added
- Initial implementation of the Discord bot.
- Integration with the `discord.ext` module for creating a bot instance.
- Integration with `configparser` for reading the bot's configuration from `config.ini`.
- Integration with OpenAI API for generating AI responses using ChatGPT 3.5 Turbo.
- Implementation of a cache mechanism to store and retrieve generated responses.
- Implementation of a function to generate AI responses based on user messages.
- Event handler for the `on_message` event to process user messages.
- Handling of bot messages by deleting them after 2 minutes.
- Handling of user messages in the "ask-anything" channel.
- Checking user roles for special access and verification.
- Handling of access denial for users without the required roles.
- Handling of API errors and other exceptions during message processing.
- Event handler for the `on_ready` event to print bot readiness information.
- Event handler for command-related errors.
- Starting the bot by running the `bot.run()` function.

## CBT3.5 - ChatBotTurbo3.5 V1- 2023-07-13

### Added
- Initial release of the Discord Chat Bot.


ChatBotTurbo3.5/main.py
import discord
from discord.ext import commands
import configparser
import openai
import asyncio

# Read config.ini
config = configparser.ConfigParser()
config.read('config.ini')

token = config['Bot']['token']
api_key = config['OpenAI']['api_key']

# Create an instance of the commands.Bot class
intents = discord.Intents.default()
intents.typing = False
intents.presences = False
intents.message_content = True

bot = commands.Bot(command_prefix='!', intents=intents)

# Initialize the OpenAI API
openai.api_key = api_key

# Define a dictionary to cache generated responses
response_cache = {}

# Define a function to generate a response using ChatGPT 3.5 Turbo
async def generate_response(message):
    model_prompt = f"User: {message.content}\nAI: "
    cache_key = model_prompt + str(message.author.id)
    
    if cache_key in response_cache:
        generated_text = response_cache[cache_key]
    else:
        response = openai.Completion.create(
            engine="text-davinci-003",
            prompt=model_prompt,
            max_tokens=100,
            temperature=0.7,
            n=1,
            stop=None,
            user=str(message.author.id),
        )
        generated_text = response.choices[0].text.strip()
        
        response_cache[cache_key] = generated_text
    
    if "?" not in generated_text:
        prompt_with_question = f"{generated_text} AI: "
        cache_key = prompt_with_question + str(message.author.id)
        
        if cache_key in response_cache:
            generated_text = response_cache[cache_key]
        else:
            response = openai.Completion.create(
                engine="text-davinci-003",
                prompt=prompt_with_question,
                max_tokens=100,
                temperature=0.7,
                n=1,
                stop=None,
                user=str(message.author.id),
            )
            generated_text = response.choices[0].text.strip()
            
            response_cache[cache_key] = generated_text
    
    return generated_text

# Event handler for the on_message event
@bot.event
async def on_message(message):
    if message.author == bot.user:
        # Delete bot messages after 2 minutes
        await asyncio.sleep(120)
        try:
            await message.delete()
        except discord.errors.NotFound:
            pass
        return

    if message.channel.name == "ask-anything":
        special_access_role = discord.utils.get(message.guild.roles, name="Special Access")
        verified_role = discord.utils.get(message.guild.roles, name="Verified")

        if special_access_role in message.author.roles and verified_role in message.author.roles:
            try:
                response = await generate_response(message)
                await message.channel.send(response)

                # Delete user question after 1 minute
                await asyncio.sleep(60)
                try:
                    await message.delete()
                except discord.errors.NotFound:
                    pass

                # Delete bot response after 2 minutes
                await asyncio.sleep(120)
                async for msg in message.channel.history(limit=200):
                    if msg.author == bot.user and msg.reference is not None and msg.reference.message_id == message.id:
                        try:
                            await msg.delete()
                        except discord.errors.NotFound:
                            pass
                        break
            except openai.Error as e:
                # Handle OpenAI API errors
                await message.channel.send(f"OpenAI API error: {str(e)}")
            except Exception as e:
                # Handle other errors
                await message.channel.send(f"An error occurred: {str(e)}")
        elif verified_role in message.author.roles:
            # User has only "Verified" role
            # Delete user message instantly
            try:
                await message.delete()
            except discord.errors.NotFound:
                pass

            # Send access denial message if not already sent
            error_sent = False
            async for msg in message.channel.history(limit=10):
                if msg.author == bot.user and msg.content == "You do not have the required role for special access.":
                    error_sent = True
                    break
            if not error_sent:
                await message.channel.send("You do not have the required role for special access.")

        else:
            # Send access denial message if not already sent
            error_sent = False
            async for msg in message.channel.history(limit=10):
                if msg.author == bot.user and msg.content == "You do not have the required role for special access.":
                    error_sent = True
                    break
            if not error_sent:
                await message.channel.send("You do not have the required role for special access.")

    await bot.process_commands(message)

# Event handler for the on_ready event
@bot.event
async def on_ready():
    print(f'Bot is ready. Connected to {len(bot.guilds)} server(s).')

# Event handler for command-related errors
@bot.event
async def on_command_error(ctx, error):
    if isinstance(error, commands.CommandNotFound):
        return  # Ignore invalid commands silently
    elif isinstance(error, commands.MissingRequiredArgument):
        await ctx.send("Missing required argument.")
    elif isinstance(error, openai.Error):
        await ctx.send(f"OpenAI API error: {str(error)}")
    else:
        # Log the error and send a generic error message
        error_message = f"An error occurred: {str(error)}"
        print(f"Error: {error_message}")
        await ctx.send("An error occurred while processing the command.")

# Start the bot
bot.run(token)
